### PR TITLE
Foward looking gazebo sonar initial commit

### PR DESCRIPTION
**Pull request by ivan1993br**
Initial commit of the gazebo ros sonar based on the shader code for rock.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,105 @@
-# CMakeLists.txt has to be located in the project folder and cmake has to be
-# executed from 'project/build' with 'cmake ../'.
-cmake_minimum_required(VERSION 2.6)
-find_package(Rock)
-rock_init(vizkit3d_normal_depth_map 0.1)
-rock_standard_layout()
+cmake_minimum_required(VERSION 2.8.3)
+project(gazebo_ros_sonar_plugin)
```
Comment (g-arjones) [inline CMakeLists.txt]: Why are the repository name and project name different? Some tools assume they are always the same...

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,6 +1,105 @@
...
+  ${GAZEBO_ROS_SONAR_PLUGIN_HEADERS})
+
+add_library(FLSonar ${GAZEBO_SONAR_PLUGIN_SRCS})
+target_link_libraries(FLSonar ${GAZEBO_LIBRARIES} ${OpenCV_LIBRARIES})
+list(APPEND GAZEBO_SONAR_PLUGIN_LIST FLSonar)
+
+add_library(fl_sonar_ros ${GAZEBO_ROS_SONAR_PLUGIN_SRCS})
+target_link_libraries(fl_sonar_ros ${catkin_LIBRARIES} ${GAZEBO_SONAR_PLUGIN_LIST})
+add_dependencies(fl_sonar_ros ${catkin_EXPORTED_TARGETS})
+list(APPEND GAZEBO_ROS_SONAR_PLUGIN_LIST fl_sonar_ros)
```
Comment (g-arjones) [inline CMakeLists.txt]: Why does one plugin use PascalCase naming and the other one use snake_case naming? Gazebo base plugins use PascalCase, you should probably stick to that..

File: package.xml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  <depend>cv_bridge</depend>
+
+  <exec_depend>gazebo_ros</exec_depend>
+  <exec_depend>opencv3</exec_depend>
+
+  <test_depend>rosunit</test_depend>
+  <test_depend>rostest</test_depend>
+  
+  <build_depend>opencv3</build_depend>
+  <build_depend>roslint</build_depend>
```
Comment (g-arjones) [inline package.xml]: I think this package also depends on `gazebo_dev` to build...

File: src/FLSonar.cc [hunk 4]
```diff
@@ -0,0 +1,712 @@
...
+  this->camera->setFOVy( fov_now	); 
+  this->camera->setAspectRatio(1.0);
+  this->camera->setAutoAspectRatio (0);
+
+  {sdf::ElementPtr iSdf = _sdf->GetElement("clip");
+  GZ_ASSERT(iSdf!=nullptr,"clip is not set");
+  GZ_ASSERT(iSdf->Get<double>("far"),"far element of clip is not set");	
+  this->SetFarClip(iSdf->Get<double>("far"));}
+
+  {sdf::ElementPtr iSdf = _sdf->GetElement("clip");
```
Comment (g-arjones) [inline src/FLSonar.cc]: Why are you asserting this twice?

File: src/FLSonar.cc [hunk 5]
```diff
@@ -0,0 +1,712 @@
...
+  GZ_ASSERT(iSdf!=nullptr,"clip is not set");
+  GZ_ASSERT(iSdf->Get<double>("near"),"near element of clip is not set");	
+  this->SetNearClip(iSdf->Get<double>("near"));}
+
+  {sdf::ElementPtr iSdf = _sdf->GetElement("image");
+  GZ_ASSERT(iSdf!=nullptr,"Image is not set");
+  GZ_ASSERT(iSdf->Get<double>("width"),"Width element of image is not set");	
+  this->SetImageWidth(iSdf->Get<double>("width"));}
+
+  {sdf::ElementPtr iSdf = _sdf->GetElement("image");
```
Comment (g-arjones) [inline src/FLSonar.cc]: Why are you asserting this twice?

File: src/FLSonar.cc [hunk 6]
```diff
@@ -0,0 +1,712 @@
...
+
+  Ogre::Pass *pass = technique->getPass(0);
+  GZ_ASSERT(pass, "Sonar material script error: pass not found");
+  GZ_ASSERT(pass->hasVertexProgram(),"Must have vertex program");
+  GZ_ASSERT(pass->hasFragmentProgram(),"Must have vertex program");}
+
+}
+
+//////////////////////////////////////////////////
+void FLSonar::ImageTextureToCV(float _width,int _height, Ogre::Texture* _inTex)
```
Comment (g-arjones) [inline src/FLSonar.cc]: There are 23 lines in this method, only 7 are actually meaningful, the rest is either commented out or empty lines. This makes reading your code worse than it should be...

File: src/FLSonar.cc [hunk 7]
```diff
@@ -0,0 +1,712 @@
...
+  if (pass->hasFragmentProgram())
+  {
+    renderSys->bindGpuProgram(
+    pass->getFragmentProgram()->_getBindingDelegate());
+
+#if OGRE_VERSION_MAJOR == 1 && OGRE_VERSION_MINOR == 6
+    renderSys->bindGpuProgramParameters(Ogre::GPT_FRAGMENT_PROGRAM,
+    pass->getFragmentProgramParameters());
+#else
+      renderSys->bindGpuProgramParameters(Ogre::GPT_FRAGMENT_PROGRAM,
```
Comment (g-arjones) [inline src/FLSonar.cc]: Indentation

File: src/FLSonarRos.cc [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+      this->scene = rendering::RenderEngine::Instance()->CreateScene(worldName,false,true);
+      gzwarn << "Que budega" << std::endl;
+    }
+
+    if(scene != nullptr)
+    { 
+      
+      gzwarn << "Got Scene" << std::endl;
+      double hfov = M_PI/2;
+      this->sonar=boost::shared_ptr<rendering::FLSonar>(new rendering::FLSonar(this->sensor->Name(),this->scene,false));
```
Comment (g-arjones) [inline src/FLSonarRos.cc]: You are using std::bind in your code, which makes your code require C++11 to compile. Why use boost::shared_ptr over std::shared_ptr in this case?

File: xacro/default_car_sonar.xacro [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  <xacro:arg name="debug" default="0"/>
+  <xacro:arg name="namespace" default="car_sonar"/>
+  <xacro:arg name="concept" default="car_sonar"/>
+  <!-- -->
+  <!-- Include the ROV macro file -->
+  <xacro:include filename="$(find gazebo_ros_sonar_plugin)/xacro/base_$(arg concept).xacro"/>
+  <xacro:include filename="$(find gazebo_ros_sonar_plugin)/xacro/gazebo_$(arg concept).xacro"/>
+  <!-- -->
+  <!-- Create the mccr -->
+  <xacro:mccr_base namespace="$(arg namespace)">
```
Comment (g-arjones) [inline xacro/default_car_sonar.xacro]: mccr?

File: CMakeLists.txt (hunk 1 above)
Comment (ivan1993br) [reply CMakeLists.txt]: Ok

File: src/FLSonar.cc (hunk 6 above)
Comment (ivan1993br) [reply src/FLSonar.cc]: Ok

File: src/FLSonar.cc (hunk 7 above)
Comment (ivan1993br) [reply src/FLSonar.cc]: Ok

File: src/FLSonarRos.cc (hunk 8 above)
Comment (ivan1993br) [reply src/FLSonarRos.cc]: Ok

File: xacro/default_car_sonar.xacro (hunk 9 above)
Comment (ivan1993br) [reply xacro/default_car_sonar.xacro]: Yes, my bad

File: package.xml (hunk 3 above)
Comment (ivan1993br) [reply package.xml]: Yes, fixed

**Comment by ivan1993br** [comment]
Sorry about the time for correction, but I made the modification proposed by Gabriel, linted the code, updated the dependencies and changed the name.

File: src/FLSonarRos.cc [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+  this->scene = rendering::get_scene(worldName);
+
+  // Get scene pointer
+  gzwarn << rendering::RenderEngine::Instance()->SceneCount() << " Num Scenes" << std::endl;
+
+  if (!this->scene)
+  {
+    this->scene = rendering::RenderEngine::Instance()->CreateScene(worldName, false, true);
+    gzwarn << "Que budega" << std::endl;
```
Comment (g-arjones) [inline src/FLSonarRos.cc]: ??

File: src/FLSonar.cc [hunk 11]
```diff
@@ -0,0 +1,689 @@
...
+
+    GZ_ASSERT(iSdf->Get<double>("height"), "Height element of image is not set");
+    this->SetImageHeight(iSdf->Get<double>("height"));
+  }
+
+  GZ_ASSERT(_sdf->Get<double>("bin_count"), "bin_count is not set");
+  this->SetBinCount(_sdf->Get<double>("bin_count"));
+
+  GZ_ASSERT(_sdf->Get<double>("beam_count"), "beam_count is not set");
+  this->SetBeamCount(_sdf->Get<double>("beam_count"));
```
Comment (g-arjones) [inline src/FLSonar.cc]: This stuff still doesn't look good. Why don't you factor this out in a helper static local function that does the assertion and returns the value of the element? Would look much better

File: src/FLSonarRos.cc (hunk 10 above)
Comment (ivan1993br) [reply src/FLSonarRos.cc]: My bad...

File: src/FLSonar.cc (hunk 11 above)
Comment (ivan1993br) [reply src/FLSonar.cc]: Changed

File: src/FLSonar.cc [hunk 12]
```diff
@@ -0,0 +1,689 @@
...
+#include "gazebo/common/Mesh.hh"
+#include "gazebo/common/MeshManager.hh"
+#include "gazebo/common/Timer.hh"
+
+#include "gazebo/rendering/Visual.hh"
+#include "gazebo/rendering/Conversions.hh"
+#include "gazebo/rendering/Scene.hh"
+#include "foward_looking_gazebo_sonar/FLSonar.hh"
+
+using namespace gazebo;  // NOLINT(build/namespaces)
```
[the remark this answers is not in the document]
Comment (ivan1993br) [reply src/FLSonar.cc]: Fixed

File: src/FLSonar.cc [hunk 13]
```diff
@@ -0,0 +1,695 @@
...
+    this->SetImageHeight(this->GetSDFElement(iSdf, "height"));
+  }
+
+  this->SetBinCount(this->GetSDFElement(_sdf, "bin_count"));
+
+  this->SetBeamCount(this->GetSDFElement(_sdf, "beam_count"));
+}
+
+//////////////////////////////////////////////////
+double FLSonar::GetSDFElement(sdf::ElementPtr &_sdf, const std::string _nameElement)
```
Comment (g-arjones) [inline src/FLSonar.cc]: Ok.. You are almost there. How about:

```C++
double FLSonar::GetSDFElement(const sdf::ElementPtr &_sdf, const std::string &_nameElement,
                              const std::string &_parent = "")
{
  sdf::ElementPtr parentSdf = _sdf;
  if (!_parent.empty())
  {
    parentSdf = _sdf->GetElement(_parent);
    GZ_ASSERT(parentSdf != nullptr, (_parent + " is not set").c_str());
  }

  GZ_ASSERT(parentSdf->Get<double>(_nameElement), std::string(_nameElement + " is not set").c_str());
  return parentSdf->Get<double>(_nameElement);
}
```

Then your load method becomes:
```C++
void FLSonar::Load(sdf::ElementPtr _sdf)
{
  Camera::Load(_sdf);
  this->SetVertFOV(this->GetSDFElement(_sdf, "vfov"));
  
  Ogre::Radian fov_now(this->vfov);
  this->camera->setFOVy(fov_now);
  this->camera->setAspectRatio(1.0);
  this->camera->setAutoAspectRatio(0);

  this->SetFarClip(this->GetSDFElement(_sdf, "far", "clip"));
  this->SetNearClip(this->GetSDFElement(_sdf, "near", "clip"));
  this->SetImageWidth(this->GetSDFElement(_sdf, "width", "image"));
  this->SetImageHeight(this->GetSDFElement(_sdf, "height", "image"));
  this->SetBinCount(this->GetSDFElement(_sdf, "bin_count"));
  this->SetBeamCount(this->GetSDFElement(_sdf, "beam_count"));
}
```

File: CMakeLists.txt [hunk 14]
```diff
@@ -1,6 +1,105 @@
...
+install(DIRECTORY include/${PROJECT_NAME}/
+  DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION}
+  FILES_MATCHING PATTERN "*.hh"
+  PATTERN "*~" EXCLUDE
+)
+
+# Install urdf files
+#install(DIRECTORY urdf
+#  DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}
+#  PATTERN "*~" EXCLUDE)
```
Comment (g-arjones) [inline CMakeLists.txt]: Could you please remove this commented out code?

File: CMakeLists.txt [hunk 15]
```diff
@@ -1,6 +1,105 @@
...
+    ${GAZEBO_INCLUDE_DIRS}
+    ${TinyXML_INCLUDE_DIRS})
+
+  link_directories(
+    ${GAZEBO_LIBRARY_DIRS}
+  )
+
+  
+
+  #MESSAGE(FATAL_ERROR ${GAZEBO_LIBRARY_DIRS})
```
Comment (g-arjones) [inline CMakeLists.txt]: Could you please remove this and also the empty lines? (One is ok, more that that is unnecessary)

File: test/IncludePaths.h.in [hunk 16]
```diff
@@ -0,0 +1,12 @@
+#ifndef INCLUDE_PATHS_H
```
Comment (g-arjones) [inline test/IncludePaths.h.in]: I think you forgot to remove this file...

**Pull request merged**